geo/src/algorithm/validation/mod.rs

```diff
@@ -0,0 +1,154 @@
+///
+/// // Get the first validation error, as a `Result`
+/// let validation_error = invalid_polygon.check_validation().unwrap_err();
+/// use geo::algorithm::validation::{InvalidPolygon, RingRole};
+/// assert_eq!(validation_error, InvalidPolygon::TooFewPointsInRing(RingRole::Exterior));
+///
+/// // Get a human readable error
+/// let text = validation_error.to_string();
+/// assert_eq!(text, "exterior ring must have at least 3 distinct points");
+//
+/// // Get all validation errors
+/// let all_validation_errors = invalid_polygon.validation_errors();
+/// assert_eq!(all_validation_errors.len(), 2);
+/// assert_eq!(all_validation_errors[0].to_string(), "exterior ring must have at least 3 distinct points");
+/// assert_eq!(all_validation_errors[1].to_string(), "interior ring at index 0 is not contained within the polygon's exterior");
+/// ```
+pub trait Validation {
+    type Error: std::error::Error;
+
+    /// Check if the geometry is valid.
+    fn is_valid(&self) -> bool {
+        self.check_validation().is_ok()
+    }
+
+    /// Return the reason(s) of invalidity of the geometry.
+    ///
+    /// Though we try to return *all* problems with a geometry, it's possible that previous errors
+    /// will obscure subsequent errors. For example, a MultiPolygon requires all its elements to be
+    /// valid and non-overlapping. If one of the individual polygons is invalid, we can't guarantee
+    /// the correctness of their "overlap" check which assumes valid input. Therefore, you should
+    /// re-validate after attempting to correct any validation errors.
+    fn validation_errors(&self) -> Vec<Self::Error> {
+        let mut validation_errors = Vec::new();
+
+        self.visit_validation(Box::new(|problem| {
+            validation_errors.push(problem);
+            Ok::<(), Self::Error>(())
+        }))
+        .expect("no errors are returned");
+
+        validation_errors
+    }
+
+    /// Return the first reason of invalidity of the geometry.
+    fn check_validation(&self) -> Result<(), Self::Error> {
+        self.visit_validation(Box::new(Err))
+    }
+
+    /// Visit the validation of the geometry.
+    ///
+    /// The closure `handle_validation_error` is called for each validation error.
+    fn visit_validation<T>(
+        &self,
+        handle_validation_error: Box<dyn FnMut(Self::Error) -> Result<(), T> + '_>,
+    ) -> Result<(), T>;
+}
+
+/// The role of a ring in a [`Polygon`](crate::Polygon).
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum RingRole {
+    Exterior,
+    Interior(usize),
+}
+
+impl fmt::Display for RingRole {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            RingRole::Exterior => write!(f, "exterior ring"),
+            RingRole::Interior(idx) => write!(f, "interior ring at index {}", idx),
+        }
+    }
+}
+
+/// The position of the problem in a multi-geometry, starting at 0.
+#[derive(Debug, PartialEq, Clone)]
+pub struct GeometryIndex(usize);
```

Comment:
Is the unnamed field here intentionally not `pub`? As long as this structure is exposed publicly, it might be helpful for users to be able to read the value

Reply:
Good point - done!